capstone/backend/src/main/java/com/google/servlets/StoreRankingsServlet.java

```diff
@@ -0,0 +1,179 @@
+            userPreferences.getLocation());
+    stores =
+        stores.stream()
+            .filter(
+                store ->
+                    (distances.get(store.getStoreAddress())
+                        < (userPreferences.getDistancePreference() * MILES_TO_METERS)))
+            .collect(Collectors.toList());
+    rankStores(stores, distances);
+    response.setContentType("application/json");
+    response.setStatus(HttpServletResponse.SC_OK);
+    response.getWriter().println(g.toJson(stores));
+  }
+
+  private void rankStores(List<Store> stores, Map<String, Integer> distances) {
+    Collections.sort(
+        stores,
+        new Comparator<Store>() {
+          @Override
+          public int compare(Store s1, Store s2) {
+            double s1StoreScore =
+                s1.getNumberOfItemsFound() * AVALIABLE_ITEMS_WEIGHT
+                    + s1.getLowestPotentialPrice() * PRICE_WEIGHT
+                    + distances.get(s1.getStoreAddress()) * DISTANCE_WEIGHT;
+            double s2StoreScore =
+                s2.getNumberOfItemsFound() * AVALIABLE_ITEMS_WEIGHT
+                    + s2.getLowestPotentialPrice() * PRICE_WEIGHT
+                    + distances.get(s2.getStoreAddress()) * DISTANCE_WEIGHT;
+            if (s1StoreScore < s2StoreScore) {
+              return -1;
+            } else if (s1StoreScore > s2StoreScore) {
+              return 1;
+            } else {
+              return 0;
+            }
+          }
+        });
+  }
+
+  public Map<String, Integer> getDistances(List<String> addresses, List<Double> userLocation)
+      throws IOException {
+    HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory();
+    StringBuilder sb =
+        new StringBuilder("https://maps.googleapis.com/maps/api/distancematrix/json?origins=");
+    sb.append(userLocation.get(0) + "," + userLocation.get(1) + "&destinations=");
+    for (int i = 0; i < addresses.size(); i++) {
+      sb.append(addresses.get(i).replace(' ', '+'));
+      if (i == addresses.size() - 1) {
+        sb.append("&key=" + API_KEY);
+      } else {
+        sb.append("|");
+      }
+    }
+
+    HttpRequest request = requestFactory.buildGetRequest(new GenericUrl(sb.toString()));
+    String response = request.execute().parseAsString();
+    DistanceResponse distanceResponse = g.fromJson(response, DistanceResponse.class);
+    try {
+      return IntStream.range(0, addresses.size())
+          .boxed()
+          .collect(
+              Collectors.toMap(
+                  i -> addresses.get(i),
+                  i ->
+                      Integer.parseInt(
+                          distanceResponse.rows.get(0).elements.get(i).distance.value)));
+    } catch (NullPointerException | IndexOutOfBoundsException e) {
+      return IntStream.range(0, addresses.size())
+          .boxed()
+          .collect(Collectors.toMap(i -> addresses.get(i), i -> 0));
```

Comment:
So if one address is not found, you set 0  to all addresses?

Reply:
If an address isn't found, its because the user's location is invalid. Hence, all of the distance values would be null. The stores in the database all have valid addresses so there shouldn't be a case where some of the distances are valid and some aren't.

Reply:
Yeah, prefer not to use magic constants like 0. would rather return an empty map here.